### PR TITLE
Update to work with v0.4 as well as v0.3

### PR DESCRIPTION
**Pull request by ScottPJones**
Use better typing (Cstring instead of Ptr{Uint8}, Ptr{UChar} instead of Ptr{Uint16}
Add finalizers to avoid memory leakage
Fix some over allocation issues
Except for locale argument, always pass length instead of depend on \0 termination

### CODE REVIEW

**Comment by ScottPJones** [comment]
This package also needs unit tests for all of the functions.
@nolta We are likely to be using this heavily, are you still actively maintaining it?  If not, could it maybe be moved to JuliaLang (or possibly add me as a contributor)?  Thanks!

**Comment by nolta** [comment]
Thanks Scott. Could you split the buffer size changes into a separate pull request?

File: deps/build.jl [hunk 1]
```diff
@@ -22,7 +24,7 @@ icui18n = library_dependency("icui18n", aliases=i18n_aliases)
     using WinRPM
     provides(WinRPM.RPM, "icu", [icu,icui18n], os=:Windows)
 end
-provides(AptGet, {"libicu$v" => [icu,icui18n] for v in apt_versions})
+provides(AptGet, Dict{Any,Any}(["libicu$v" => [icu,icui18n] for v in apt_versions]))
 provides(Yum, "icu", [icu, icui18n])
 
-@BinDeps.install [:icu => :iculib, :icui18n => :iculibi18n]
+@BinDeps.install Dict([:icu => :iculib, :icui18n => :iculibi18n])
```
Comment (nolta) [inline deps/build.jl]: I think you meant `@compat Dict(:icu => ...)`?

Comment (ScottPJones) [inline deps/build.jl]: That didn't work on both 0.3 and 0.4, I'm  not sure why

**Comment by ScottPJones** [comment]
Sure, I should have thought of that, putting them in different commits or PRs, I was in a rush to make it work in 0.4 for a colleague.

File: deps/build.jl (hunk 1 above)
Comment (nolta) [inline deps/build.jl]: Hmm, seems to work for me.

**Comment by ScottPJones** [comment]
Sorry I took so long to get this updated, Labor Day weekend here in the US and all, y'know!
This doesn't have the buffer size changes, *however* I realized that it is possible that even with 2* buffer size, given the ICU rules, that for FoldCase, ToUpper, and ToTitle, you may need 3* the # of 16-bit words, as you can see by the test below:
```julia
julia> for rng in (0:0xd7ff, 0xe000:0x10ffff), c in rng
          ch = Char(c)
          str = utf16([ch]) ; sz = sizeof(str)
          low = u_strToLower(str) ; sl = sizeof(low) ; fld = u_strFoldCase(str) ; sf = sizeof(fld)
          upp = u_strToUpper(str) ; su = sizeof(upp)
          tit = u_strToTitle(str) ; st = sizeof(tit)
          if sf != sz || sl != sz || su != sz || st != sz
             println("$(hex(c)):$str\tfld:$sf $fld\tlow:$sl $low\tupp:$su $upp\ttit:$st $tit")
          end
       end
df:ß	fld:4 ss	low:2 ß	upp:4 SS	tit:4 Ss
130:İ	fld:4 i̇	low:4 i̇	upp:2 İ	tit:2 İ
149:ŉ	fld:4 ʼn	low:2 ŉ	upp:4 ʼN	tit:4 ʼN
1f0:ǰ	fld:4 ǰ	low:2 ǰ	upp:4 J̌	tit:4 J̌
390:ΐ	fld:6 ΐ	low:2 ΐ	upp:6 Ϊ́	tit:6 Ϊ́
3b0:ΰ	fld:6 ΰ	low:2 ΰ	upp:6 Ϋ́	tit:6 Ϋ́
587:և	fld:4 եւ	low:2 և	upp:4 ԵՒ	tit:4 Եւ
1e96:ẖ	fld:4 ẖ	low:2 ẖ	upp:4 H̱	tit:4 H̱
1e97:ẗ	fld:4 ẗ	low:2 ẗ	upp:4 T̈	tit:4 T̈
1e98:ẘ	fld:4 ẘ	low:2 ẘ	upp:4 W̊	tit:4 W̊
1e99:ẙ	fld:4 ẙ	low:2 ẙ	upp:4 Y̊	tit:4 Y̊
1e9a:ẚ	fld:4 aʾ	low:2 ẚ	upp:4 Aʾ	tit:4 Aʾ
1e9e:ẞ	fld:4 ss	low:2 ß	upp:2 ẞ	tit:2 ẞ
1f50:ὐ	fld:4 ὐ	low:2 ὐ	upp:4 Υ̓	tit:4 Υ̓
1f52:ὒ	fld:6 ὒ	low:2 ὒ	upp:6 Υ̓̀	tit:6 Υ̓̀
1f54:ὔ	fld:6 ὔ	low:2 ὔ	upp:6 Υ̓́	tit:6 Υ̓́
1f56:ὖ	fld:6 ὖ	low:2 ὖ	upp:6 Υ̓͂	tit:6 Υ̓͂
1f80:ᾀ	fld:4 ἀι	low:2 ᾀ	upp:4 ἈΙ	tit:2 ᾈ
1f81:ᾁ	fld:4 ἁι	low:2 ᾁ	upp:4 ἉΙ	tit:2 ᾉ
1f82:ᾂ	fld:4 ἂι	low:2 ᾂ	upp:4 ἊΙ	tit:2 ᾊ
1f83:ᾃ	fld:4 ἃι	low:2 ᾃ	upp:4 ἋΙ	tit:2 ᾋ
1f84:ᾄ	fld:4 ἄι	low:2 ᾄ	upp:4 ἌΙ	tit:2 ᾌ
1f85:ᾅ	fld:4 ἅι	low:2 ᾅ	upp:4 ἍΙ	tit:2 ᾍ
1f86:ᾆ	fld:4 ἆι	low:2 ᾆ	upp:4 ἎΙ	tit:2 ᾎ
1f87:ᾇ	fld:4 ἇι	low:2 ᾇ	upp:4 ἏΙ	tit:2 ᾏ
1f88:ᾈ	fld:4 ἀι	low:2 ᾀ	upp:4 ἈΙ	tit:2 ᾈ
1f89:ᾉ	fld:4 ἁι	low:2 ᾁ	upp:4 ἉΙ	tit:2 ᾉ
1f8a:ᾊ	fld:4 ἂι	low:2 ᾂ	upp:4 ἊΙ	tit:2 ᾊ
1f8b:ᾋ	fld:4 ἃι	low:2 ᾃ	upp:4 ἋΙ	tit:2 ᾋ
1f8c:ᾌ	fld:4 ἄι	low:2 ᾄ	upp:4 ἌΙ	tit:2 ᾌ
1f8d:ᾍ	fld:4 ἅι	low:2 ᾅ	upp:4 ἍΙ	tit:2 ᾍ
1f8e:ᾎ	fld:4 ἆι	low:2 ᾆ	upp:4 ἎΙ	tit:2 ᾎ
1f8f:ᾏ	fld:4 ἇι	low:2 ᾇ	upp:4 ἏΙ	tit:2 ᾏ
1f90:ᾐ	fld:4 ἠι	low:2 ᾐ	upp:4 ἨΙ	tit:2 ᾘ
1f91:ᾑ	fld:4 ἡι	low:2 ᾑ	upp:4 ἩΙ	tit:2 ᾙ
1f92:ᾒ	fld:4 ἢι	low:2 ᾒ	upp:4 ἪΙ	tit:2 ᾚ
1f93:ᾓ	fld:4 ἣι	low:2 ᾓ	upp:4 ἫΙ	tit:2 ᾛ
1f94:ᾔ	fld:4 ἤι	low:2 ᾔ	upp:4 ἬΙ	tit:2 ᾜ
1f95:ᾕ	fld:4 ἥι	low:2 ᾕ	upp:4 ἭΙ	tit:2 ᾝ
1f96:ᾖ	fld:4 ἦι	low:2 ᾖ	upp:4 ἮΙ	tit:2 ᾞ
1f97:ᾗ	fld:4 ἧι	low:2 ᾗ	upp:4 ἯΙ	tit:2 ᾟ
1f98:ᾘ	fld:4 ἠι	low:2 ᾐ	upp:4 ἨΙ	tit:2 ᾘ
1f99:ᾙ	fld:4 ἡι	low:2 ᾑ	upp:4 ἩΙ	tit:2 ᾙ
1f9a:ᾚ	fld:4 ἢι	low:2 ᾒ	upp:4 ἪΙ	tit:2 ᾚ
1f9b:ᾛ	fld:4 ἣι	low:2 ᾓ	upp:4 ἫΙ	tit:2 ᾛ
1f9c:ᾜ	fld:4 ἤι	low:2 ᾔ	upp:4 ἬΙ	tit:2 ᾜ
1f9d:ᾝ	fld:4 ἥι	low:2 ᾕ	upp:4 ἭΙ	tit:2 ᾝ
1f9e:ᾞ	fld:4 ἦι	low:2 ᾖ	upp:4 ἮΙ	tit:2 ᾞ
1f9f:ᾟ	fld:4 ἧι	low:2 ᾗ	upp:4 ἯΙ	tit:2 ᾟ
1fa0:ᾠ	fld:4 ὠι	low:2 ᾠ	upp:4 ὨΙ	tit:2 ᾨ
1fa1:ᾡ	fld:4 ὡι	low:2 ᾡ	upp:4 ὩΙ	tit:2 ᾩ
1fa2:ᾢ	fld:4 ὢι	low:2 ᾢ	upp:4 ὪΙ	tit:2 ᾪ
1fa3:ᾣ	fld:4 ὣι	low:2 ᾣ	upp:4 ὫΙ	tit:2 ᾫ
1fa4:ᾤ	fld:4 ὤι	low:2 ᾤ	upp:4 ὬΙ	tit:2 ᾬ
1fa5:ᾥ	fld:4 ὥι	low:2 ᾥ	upp:4 ὭΙ	tit:2 ᾭ
1fa6:ᾦ	fld:4 ὦι	low:2 ᾦ	upp:4 ὮΙ	tit:2 ᾮ
1fa7:ᾧ	fld:4 ὧι	low:2 ᾧ	upp:4 ὯΙ	tit:2 ᾯ
1fa8:ᾨ	fld:4 ὠι	low:2 ᾠ	upp:4 ὨΙ	tit:2 ᾨ
1fa9:ᾩ	fld:4 ὡι	low:2 ᾡ	upp:4 ὩΙ	tit:2 ᾩ
1faa:ᾪ	fld:4 ὢι	low:2 ᾢ	upp:4 ὪΙ	tit:2 ᾪ
1fab:ᾫ	fld:4 ὣι	low:2 ᾣ	upp:4 ὫΙ	tit:2 ᾫ
1fac:ᾬ	fld:4 ὤι	low:2 ᾤ	upp:4 ὬΙ	tit:2 ᾬ
1fad:ᾭ	fld:4 ὥι	low:2 ᾥ	upp:4 ὭΙ	tit:2 ᾭ
1fae:ᾮ	fld:4 ὦι	low:2 ᾦ	upp:4 ὮΙ	tit:2 ᾮ
1faf:ᾯ	fld:4 ὧι	low:2 ᾧ	upp:4 ὯΙ	tit:2 ᾯ
1fb2:ᾲ	fld:4 ὰι	low:2 ᾲ	upp:4 ᾺΙ	tit:4 Ὰͅ
1fb3:ᾳ	fld:4 αι	low:2 ᾳ	upp:4 ΑΙ	tit:2 ᾼ
1fb4:ᾴ	fld:4 άι	low:2 ᾴ	upp:4 ΆΙ	tit:4 Άͅ
1fb6:ᾶ	fld:4 ᾶ	low:2 ᾶ	upp:4 Α͂	tit:4 Α͂
1fb7:ᾷ	fld:6 ᾶι	low:2 ᾷ	upp:6 Α͂Ι	tit:6 ᾼ͂
1fbc:ᾼ	fld:4 αι	low:2 ᾳ	upp:4 ΑΙ	tit:2 ᾼ
1fc2:ῂ	fld:4 ὴι	low:2 ῂ	upp:4 ῊΙ	tit:4 Ὴͅ
1fc3:ῃ	fld:4 ηι	low:2 ῃ	upp:4 ΗΙ	tit:2 ῌ
1fc4:ῄ	fld:4 ήι	low:2 ῄ	upp:4 ΉΙ	tit:4 Ήͅ
1fc6:ῆ	fld:4 ῆ	low:2 ῆ	upp:4 Η͂	tit:4 Η͂
1fc7:ῇ	fld:6 ῆι	low:2 ῇ	upp:6 Η͂Ι	tit:6 ῌ͂
1fcc:ῌ	fld:4 ηι	low:2 ῃ	upp:4 ΗΙ	tit:2 ῌ
1fd2:ῒ	fld:6 ῒ	low:2 ῒ	upp:6 Ϊ̀	tit:6 Ϊ̀
1fd3:ΐ	fld:6 ΐ	low:2 ΐ	upp:6 Ϊ́	tit:6 Ϊ́
1fd6:ῖ	fld:4 ῖ	low:2 ῖ	upp:4 Ι͂	tit:4 Ι͂
1fd7:ῗ	fld:6 ῗ	low:2 ῗ	upp:6 Ϊ͂	tit:6 Ϊ͂
1fe2:ῢ	fld:6 ῢ	low:2 ῢ	upp:6 Ϋ̀	tit:6 Ϋ̀
1fe3:ΰ	fld:6 ΰ	low:2 ΰ	upp:6 Ϋ́	tit:6 Ϋ́
1fe4:ῤ	fld:4 ῤ	low:2 ῤ	upp:4 Ρ̓	tit:4 Ρ̓
1fe6:ῦ	fld:4 ῦ	low:2 ῦ	upp:4 Υ͂	tit:4 Υ͂
1fe7:ῧ	fld:6 ῧ	low:2 ῧ	upp:6 Ϋ͂	tit:6 Ϋ͂
1ff2:ῲ	fld:4 ὼι	low:2 ῲ	upp:4 ῺΙ	tit:4 Ὼͅ
1ff3:ῳ	fld:4 ωι	low:2 ῳ	upp:4 ΩΙ	tit:2 ῼ
1ff4:ῴ	fld:4 ώι	low:2 ῴ	upp:4 ΏΙ	tit:4 Ώͅ
1ff6:ῶ	fld:4 ῶ	low:2 ῶ	upp:4 Ω͂	tit:4 Ω͂
1ff7:ῷ	fld:6 ῶι	low:2 ῷ	upp:6 Ω͂Ι	tit:6 ῼ͂
1ffc:ῼ	fld:4 ωι	low:2 ῳ	upp:4 ΩΙ	tit:2 ῼ
fb00:ﬀ	fld:4 ff	low:2 ﬀ	upp:4 FF	tit:4 Ff
fb01:ﬁ	fld:4 fi	low:2 ﬁ	upp:4 FI	tit:4 Fi
fb02:ﬂ	fld:4 fl	low:2 ﬂ	upp:4 FL	tit:4 Fl
fb03:ﬃ	fld:6 ffi	low:2 ﬃ	upp:6 FFI	tit:6 Ffi
fb04:ﬄ	fld:6 ffl	low:2 ﬄ	upp:6 FFL	tit:6 Ffl
fb05:ﬅ	fld:4 st	low:2 ﬅ	upp:4 ST	tit:4 St
fb06:ﬆ	fld:4 st	low:2 ﬆ	upp:4 ST	tit:4 St
fb13:ﬓ	fld:4 մն	low:2 ﬓ	upp:4 ՄՆ	tit:4 Մն
fb14:ﬔ	fld:4 մե	low:2 ﬔ	upp:4 ՄԵ	tit:4 Մե
fb15:ﬕ	fld:4 մի	low:2 ﬕ	upp:4 ՄԻ	tit:4 Մի
fb16:ﬖ	fld:4 վն	low:2 ﬖ	upp:4 ՎՆ	tit:4 Վն
fb17:ﬗ	fld:4 մխ	low:2 ﬗ	upp:4 ՄԽ	tit:4 Մխ
```

**Comment by ScottPJones** [comment]
Thanks for merging!  I wanted to know, what plans you had for the ICU package?  I'd like to continue to extend it, to support all of the ICU functionality, but if so, it might be better to organize it along the lines of the ICU prefixes, and the different ICU header files (i.e. ucal.h, ucnv.h, etc).
I'd like to make sure that whatever I do fits with any future plans.  Thanks again!

**Comment by ScottPJones** [comment]
@nolta What needs to be done so that the new version with the fixes are picked up when somebody does Pkg.add("ICU")?  Thanks!

**Pull request merged**